models/stock_picking_batch.py

```diff
@@ -0,0 +1,144 @@
+        batch = None
+
+        if not batches:
+            batch = self._create_batch(user_id)
+        else:
+            if len(batches) > 1:
+                raise ValidationError(
+                    _("Expected single picking batch, found %d batches")
+                    % len(batches))
+
+            batch = batches[0]
+
+        return batch
+
+    def _get_single_batch_info(self, batch, allowed_picking_states):
+        filtered_pickings = batch.picking_ids.filtered(
+            lambda x: x.state in allowed_picking_states)
+
+        return {'id': self.id,
+                'picking_ids': filtered_pickings.get_info()}
+
+    def get_info(self, allowed_picking_states=None):
+        """
+        Return list of dictionaries containing information about
+        all batches.
+        """
+        if allowed_picking_states is None:
+            allowed_picking_states = ['assigned']
+
+        return [self._get_single_batch_info(batch, allowed_picking_states)
+                for batch in self]
+
+    @api.multi
+    def create_batch(self, picking_priorities, user_id=None):
+        """
+        Creeate and return a batch for the specified user if pickings
+        exist. Return None otherwise. Pickings are filtered based on
+        the specified picking priorities.
+
+        Raise a ValidationError in case the user ID is not specified.
+
+        If the user already has batches assigned, a ValidationError
+        is raised in case of pickings that need to be completed,
+        otherwise such batches will be marked as done.
+        """
+        user_id = self._check_user_id(user_id)
+        self._check_batches(user_id)
+
+        return self._create_batch(user_id, picking_priorities)
+
+    def _create_batch(self, user_id, picking_priorities=None):
+        Picking = self.env['stock.picking']
+        PickingBatch = self.env['stock.picking.batch']
+        Users = self.env['res.users']
+
+        warehouse = Users.get_user_warehouse()
+        picking_type_id = warehouse.pick_type_id
+        search_domain = []
+
+        if picking_priorities is not None:
+            search_domain.append(('priority', 'in', picking_priorities))
+
+        search_domain.extend([('picking_type_id', '=', picking_type_id.id),
+                              ('state', '=', 'assigned'),
+                              ('batch_id', '=', False)])
+
+        picking = Picking.search(search_domain,
+                                 order='priority desc, scheduled_date, id',
+                                 limit=1)
+
+        if not picking:
+            return None
+
+        batch = PickingBatch.create({'user_id': user_id})
+        picking.batch_id = batch.id
+        batch.write({'state': 'in_progress'})
+
+        return batch
+
+    def _check_batches(self, user_id):
+        Picking = self.env['stock.picking']
+        PickingBatch = self.env['stock.picking.batch']
+
+        batches = PickingBatch.search([('user_id', '=', user_id),
```

Comment:
We have this search multiple times, is it worth to abstract it into a function?

Reply:
I'd rather not, for readability. Have the plain query tells you a lot right away compared to having to look for a helper.